### ISSUE TITLE
Use actual Kotlin annotation class to provide metadata to rules

### ISSUE THREAD

**Issue by chao2zhang** [detekt/detekt#3562]
We currently have `@configuration`, `@active`, and `@requiresTypeResolution` inside comment to provide metadata, and we are about to add `@controvertial` according to https://github.com/detekt/detekt/issues/3560.

Annotating comments to annotate rules is acceptable, but to further improve the quality of these metadata, we should use Kotlin annotation class to enforce information, such as `@Active(since = "1.16.0")` or `@Configuration(type = String::class)`. 

The benefits:
- Improve code review quality since we do not need to ask contributors to annotate according to a given format
- Facilitate custom detekt rules to check implementations, such as enforcing `@AutoCorrectible` rules to invoke mutation methods. (Technically `@requiresTypeResolution` should be gone in https://github.com/detekt/detekt/pull/3127/)
- Allow clients, such as gradle plugins, to configure rules based on the annotation (This PR is probably part of https://github.com/detekt/detekt/pull/3002)

**Comment by schalkms** [comment]
That's a good idea. The generated stays the same. The rules as well as the generator have to be changed.

**Comment by cortinico** [comment]
+1 for me. Does it also makes sense to expose them as part of our public API? 
Users that interact with out tools could filter or get more insights on rule metadata directly from the code.

**Comment by chao2zhang** [comment]
> Does it also makes sense to expose them as part of our public API?
Definitely.

**Comment by marschwar** [comment]
I also really like the idea. Are you already working on it? Were you also thinking about generating code from those annotations? For example it should be fairly simple to generate a type safe configuration class from the annotations as well. For example:

```kotlin
@Rule(
    since = "1.1.0",
    activeByDefault = true,
    activeByDefaultSince = "1.15.0",
    config = Config(
        ints = [
            IntConfig("threshold", defaultValue = DEFAULT_DUPLICATION),
        ],
        strings = [
            StringConfig("ignoreStringsRegex", defaultValue = IGNORE_STRINGS_REGEX_DEFAULT),
        ],
        booleans = [
            BooleanConfig("ignoreAnnotation", defaultValue = true),
            BooleanConfig("excludeStringsWithLessThan5Characters", defaultValue = true),
        ]
    )
)
class StringLiteralDuplication() ...
```

could be used to generate 

```kotlin
data class StringLiteralDuplicationConfig(
    val threshold: Int = 3,
    val ignoreStringsRegex: String = "$^",
    val ignoreAnnotation: Boolean = true,
    val excludeStringsWithLessThan5Characters: Boolean = true,
) {
    companion object {
        fun fromConfig(config: Config): StringLiteralDuplicationConfig {
            return with(config) {
                StringLiteralDuplicationConfig(
                    threshold = valueOrDefault("threshold", 3),
                    ignoreStringsRegex = valueOrDefault("ignoreStringsRegex", "$^"),
                    ignoreAnnotation = valueOrDefault("ignoreAnnotation", true),
                    excludeStringsWithLessThan5Characters = valueOrDefault(
                        "excludeStringsWithLessThan5Characters",
                        true
                    ),
                )
            }
        }
    }
}
```
This way all those config property names and their default values could become consistent in configuration, documentation and code. 

I don't want to step on anyone's toes and I think the example annotations are a bit clumsy but I am interested if annotation processing and code generation would be considered an option in detekt. I am also not sure if it meshes well with all the other things you guys are working on at the moment.

**Comment by chao2zhang** [comment]
No I am not working on it. Feel free to pick it up by assigning the PRs.

I also believe that this may require mutilple iterations, such as to add annotation for type resolution first, then for active, then for configuration.

My intuition is to avoid nesting to make the parsing easier. So I would prefer to have @Active(since = "1.16.0") @TypeResolving @Config instead of putting them under the same @Rule.

**Comment by BraisGabin** [comment]
I really like this idea!

> This way all those config property names and their default values could become consistent in configuration, documentation and code.

Anything in this direction would be great. But do we need a unique annotation to accomplish this? I'm with @chao2zhang here, multiple annotations seems like a nicer API.

**Comment by marschwar** [comment]
Of course there is no need for a single annotation. I deliberately used it
* to spark a discussion about that part ;)
* and because I thought it might be easier for rule authors to identify what is needed when a new rule is added.

I like the idea of going one annotation at a time. I would like to pick this up and start with `@RequiresTypeResolution`.

**Comment by schalkms** [comment]
I like where this discussion is going. Yes, I agree that having smaller actionable items is the key to success here. Improving the API step by step is important here, IMHO.

FYI: Briefly, ideas like this have been brought up in #1415, but I think the annotation classes are a better solution here. Hence, I'm closing #1415.

**Comment by schalkms** [comment]
As mentioned in #3568 answering the following question in the docs is beneficial for detekt's user base. Thanks for the feedback @rock3r 

> Is the rule on by default?

**Comment by marschwar** [comment]
Before I get started on the most complex kdoc tag, I like your opinion. 

I see two general ways for the configuration annotation(s).

### (1) One annotation for all types
#### Declaration
```kotlin
annotation class Configuration(
    val name: String,
    val type: ConfigurationType,
    val description: String,
    val defaultInt: Int = 0,
    val defaultString: String = "",
    val defaultBoolean: Boolean = false,
    val defaultStringArray: Array<String> = [],
)

enum class ConfigurationType {
    STRING, STRING_ARRAY, INT, BOOLEAN
}
```
#### Usage
```kotlin
@Configuration(
    name = "threshold",
    type = INT,
    description = "McCabe's Cyclomatic Complexity (MCC) number for a method",
    defaultInt = DEFAULT_THRESHOLD_METHOD_COMPLEXITY
)
@Configuration(
    name = "ignoreSingleWhenExpression",
    type = BOOLEAN,
    description = "Ignores a complex method if it only contains a single when expression",
    defaultBoolean = false
)
```

### (2) One annotation per type
#### Declaration
```kotlin
annotation class IntConfigParam(
    val name: String,
    val description: String,
    val defaultValue: Int,
)
annotation class BooleanConfigParam(
    val name: String,
    val description: String,
    val defaultValue: Boolean,
)
```

#### Usage
```kotlin
@IntConfigParam(
    name = "threshold",
    description = "McCabe's Cyclomatic Complexity (MCC) number for a method",
    defaultValue = DEFAULT_THRESHOLD_METHOD_COMPLEXITY
)
@BooleanConfigParam(
    name = "ignoreSingleWhenExpression",
    description = "Ignores a complex method if it only contains a single when expression",
    defaultValue = false
)
```

I am leaning heavily towards the second option as it communicates much clearer how it is supposed to be used. Can anyone think of a third option?

**Comment by chao2zhang** [comment]
I am wondering if we could add the annotation to Rule class properties, so that we do not need to provide additional typing.

The idea is from `https://github.com/detekt/detekt/blob/main/detekt-cli/src/main/kotlin/io/gitlab/arturbosch/detekt/cli/CliArgs.kt`

**Comment by marschwar** [comment]
That would be a very elegant solution but also a pretty big change.

**Comment by cortinico** [comment]
> I am wondering if we could add the annotation to Rule class properties, so that we do not need to provide additional typing.

+1 for this as Type and Name can be inferred then.

**Comment by marschwar** [comment]
Ok, sounds good to me if everyone is on board with public mutable fields for all rules. 
I will implement this as an optional way to specify the configuration so we do not have to change all rules in one PR.

**Comment by BraisGabin** [comment]
In 2.0 they would be immutable private fields. But that would work too, right?

**Comment by marschwar** [comment]
Never mind. It can be done immedeately

**Comment by marschwar** [comment]
I think this can be closed. If there is an annotation missing, we should create a new issue.

**Comment by BraisGabin** [comment]
Great work! Now we need to move all the old configuration to annotations
